examples/documentation/drop_operation_state_documentation.cpp

```diff
@@ -0,0 +1,34 @@
+//  Copyright (c) 2024 ETH Zurich
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+#include <pika/execution.hpp>
+#include <pika/init.hpp>
+
+#include <assert.h>
+#include <utility>
+
+int main(int argc, char* argv[])
+{
+    namespace ex = pika::execution::experimental;
+    namespace tt = pika::this_thread::experimental;
+
+    pika::start(argc, argv);
+    ex::thread_pool_scheduler sched{};
+
+    auto sp = std::make_shared<int>(42);
+    std::weak_ptr<int> sp_weak = sp;
+
+    auto s = ex::just(std::move(sp)) | ex::then([&](auto&&) { assert(sp_weak.use_count() == 1); }) |
+        ex::drop_operation_state() |
+        // If drop_operation state was not used, the shared pointer would stay alive
+        ex::then([&] { assert(sp_weak.use_count() == 0); });
```

Comment:
I realize the original test where this comes from is missing at least one more check, and I think this would be useful to add at least to the unit test, but I'll suggest it here:
```suggestion
    auto s = ex::just(std::move(sp)) |
        ex::then([&](auto&&) { assert(sp_weak.use_count() == 1); }) |
        // Even though the shared_ptr is no longer in use, it may be kept alive by the operation state
        ex::then([&]() { assert(sp_weak.use_count() == 1); }) |
        ex::drop_operation_state() |
        // Once drop_operation_state has been used, the shared_ptr is guaranteed to be released
        ex::then([&] { assert(sp_weak.use_count() == 0); });
```

I don't know if this is better, so please comment what you think: the example may also be rewritten with comments and asserts replaced by print statements saying something equivalent. I'm not sure which is easier for a user to understand. The shared_ptr may also be better demonstrated with a custom class that e.g. prints something in the destructor, but again, I'm not sure what's easiest for a new user to follow. At least shared_ptr is something that's in the standard and its behaviour is well documented.

Reply:
I separated the unit test change into #1293.

Reply:
I'm also in-between. I do feel like showing the absence of `drop_operation_state` helps demonstrate what it's trying to achieve, because it's not always obvious that the operation state/values that are no longer used are _not_ released early.

This is very related: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2024/p3373r0.pdf. But I'm not sure including it in the documentation helps explain the current situation, since that proposal may or may not end up changing the standard eventually.

Really I'd like to hear from @albestro, @rasolca, and @RMeli as users what they find explains `drop_operation_state` best.

Reply:
I think the suggested change makes thing clearer, at least to me.